core/synch.c

```diff
@@ -2289,3 +2289,201 @@ detach_on_permanent_stack(bool internal, bool do_cleanup, dr_stats_t *drstats)
+             * app that assumes no signals and assumes its non-auto-restart syscalls
+             * don't need loops could be broken.
+             */
+            LOG(GLOBAL, LOG_ALL, 3,
+                /* Having the code bytes can help diagnose post-detach where the code
+                 * cache is gone.
+                 */
+                "Detach: pre-xl8 pc=%p (%02x %02x %02x %02x %02x), xsp=%p "
+                "for thread " TIDFMT "\n",
+                mc.pc, *mc.pc, *(mc.pc + 1), *(mc.pc + 2), *(mc.pc + 3), *(mc.pc + 4),
+                mc.xsp, threads[i]->id);
+            DEBUG_DECLARE(ok =)
+            translate_mcontext(threads[i], &mc, true /*restore mem*/, NULL /*f*/);
+            ASSERT(ok);
+            if (!threads[i]->under_dynamo_control) {
+                dr_printf("Detach : thread " TIDFMT " already running natively\n",
+                          threads[i]->id);
+                LOG(GLOBAL, LOG_ALL, 1,
+                    "Detach : thread " TIDFMT " already running natively\n",
+                    threads[i]->id);
+                /* we do need to restore the app ret addr, for native_exec */
+                if (!DYNAMO_OPTION(thin_client) && DYNAMO_OPTION(native_exec) &&
+                    !vmvector_empty(native_exec_areas)) {
+                    put_back_native_retaddrs(threads[i]->dcontext);
+                }
+            }
+            LOG(GLOBAL, LOG_ALL, 1, "Detach: pc=" PFX " for thread " TIDFMT "\n", mc.pc,
+                threads[i]->id);
+            ASSERT(!is_dynamo_address(mc.pc) && !in_fcache(mc.pc));
+            /* XXX case 7457: if the thread is suspended after it received a fault
+             * but before the kernel copied the faulting context to the user mode
+             * structures for the handler, it could result in a codemod exception
+             * that wouldn't happen natively!
+             */
+            DEBUG_DECLARE(ok =)
+            thread_set_mcontext(threads[i], &mc);
+            ASSERT(ok);
+        }
+        /* Resumes the thread, which will do kernel-visible cleanup of
+         * signal state. Resume happens within the synch_all region where
+         * the thread_initexit_lock is held so that we can clean up thread
+         * data later.
+         */
+        os_signal_thread_detach(threads[i]->dcontext);
+        LOG(GLOBAL, LOG_ALL, 1, "Detach: thread " TIDFMT " is being resumed as native\n",
+            threads[i]->id);
+        os_thread_resume(threads[i]);
+    }
+    LOG(GLOBAL, LOG_ALL, 1, "Detach: waiting for threads to fully detach\n");
+    for (i = 0; i < num_threads; i++) {
+        if (i != my_idx && !IS_CLIENT_THREAD(threads[i]->dcontext))
+            os_wait_thread_detached(threads[i]->dcontext);
+    }
+    /* Clean up each thread now that everyone has gone native. Needs to be
+     * done with the thread_initexit_lock held, which is true within a synched
+     * region.
+     */
+    for (i = 0; i < num_threads; i++) {
```

Comment:
This is a large function and a lot of it looks identical to code in detach_on_permanent_stack().  Could we share at least some big chunks of it via shared helper functions?

Reply:
Thank you for your suggestion, I try to share some codes and name them `detach_set_mcontext_helper `and `detach_cleanup_helper `.